### PR TITLE
passphrase derivation + encryption + storage lib

### PR DESCRIPTION
**Pull request by ivan-zynesis**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
dependency of #344 
to serve as a scrypt protected storage, to be used to encrypt and store mnemonic seed which can be unlock using passphrase

#### Which issue(s) does this PR fixes?:

Fixes #

#### Additional comments?:
wait for #331 and #369 (merged)
this is actually worked backward, trimming down code done in 344 to ease review process

### CODE REVIEW

**Comment by fuxingloh** [comment]
Conflicts

File: packages/jellyfish-wallet-encrypted/src/scryptsy.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import scrypt from 'scryptsy'
```
Comment (ivan-zynesis) [inline packages/jellyfish-wallet-encrypted/src/scryptsy.ts:1]: Technically this is not needed, more like a implementation for `ScryptProvider` (platform agnostic) for test cases use which user (library consumer) should provide one.

The interface itself is quite difficult to tell how secured is the derived secret (from raw passphrase), this demonstrate how it is done in bip38

File: packages/jellyfish-wallet-encrypted/src/scryptsy.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  N: 16384,
+  r: 8,
+  p: 8
+}
+
+export class SimpleScryptsy implements ScryptProvider {
+  readonly scryptParams: ScryptParams
+  constructor (scryptParams?: ScryptParams) {
+    this.scryptParams = scryptParams !== undefined ? scryptParams : DEFAULT_SCRYPT_PARAMS
+  }
```
Comment (fuxingloh) [inline packages/jellyfish-wallet-encrypted/src/scryptsy.ts:20]:
```suggestion
  constructor (readonly scryptParams: ScryptParams = DEFAULT_SCRYPT_PARAMS) {
  }
```

File: packages/jellyfish-wallet-encrypted/src/scryptsy.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+  passphraseToKey (passphrase: string, salt: Buffer, keyLength: number): Buffer {
```
Comment (fuxingloh) [inline packages/jellyfish-wallet-encrypted/src/scryptsy.ts:22]: add JSDocs

File: packages/jellyfish-wallet-encrypted/src/scryptsy.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  p: number
+}
+
+const DEFAULT_SCRYPT_PARAMS: ScryptParams = {
+  N: 16384,
+  r: 8,
+  p: 8
+}
+
+export class SimpleScryptsy implements ScryptProvider {
```
Comment (fuxingloh) [inline packages/jellyfish-wallet-encrypted/src/scryptsy.ts:16]: Add JSDoc

File: packages/jellyfish-wallet-encrypted/src/scrypt-storage.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+      Number(encoded.slice(2, 4)),
+      Number(encoded.slice(4, 6)),
+      Buffer.from(encoded.slice(6, 14), 'hex'),
+      Buffer.from(encoded.slice(14, firstHalfEndIndex), 'hex'),
+      Buffer.from(encoded.slice(firstHalfEndIndex), 'hex')
+    )
+  }
+}
+
+export class ScryptStorage {
```
Comment (fuxingloh) [inline packages/jellyfish-wallet-encrypted/src/scrypt-storage.ts:64]: Add JSDoc

File: packages/jellyfish-wallet-encrypted/__tests__/scryptsy.test.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    easy.passphraseToKey('random password', Buffer.from('abcd', 'hex'), 32)
+    const easyTime = Date.now() - easyStart
+
+    const hardStart = Date.now()
+    hard.passphraseToKey('random password', Buffer.from('abcd', 'hex'), 32)
+    const hardTime = Date.now() - hardStart
+
+    // significantly slower
+    // technically it is 8x harder, but they can be processed in parallel
+    expect(hardTime).toBeGreaterThan(easyTime * 2)
```
Comment (fuxingloh) [inline packages/jellyfish-wallet-encrypted/__tests__/scryptsy.test.ts:36]: what is the `ms` for these params?

File: packages/jellyfish-wallet-encrypted/src/scrypt-storage.ts [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+export type InitVectorProvider = () => Buffer
+
+export interface Storage {
+  getter: () => Promise<string | undefined>
+  setter: (encrypted: string | undefined) => Promise<void>
+}
+
+export class EncryptedData {
```
Comment (fuxingloh) [inline packages/jellyfish-wallet-encrypted/src/scrypt-storage.ts:14]: Add JSDoc for everything, easier for me to understand

File: packages/jellyfish-wallet-encrypted/__tests__/scrypt-storage.test.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  const pass = 'password'
+
+  const data = Buffer.from(shortData, 'hex')
+  await scryptStorage.encrypt(data, pass)
+
+  const decrypted = await scryptStorage.decrypt(pass)
+  expect(decrypted?.toString('hex')).toStrictEqual(shortData)
+})
+
+it('Should reject odd number data length', async () => {
```
Comment (fuxingloh) [inline packages/jellyfish-wallet-encrypted/__tests__/scrypt-storage.test.ts:83]: odd as in `different` with supposed? Might want to rename

**Review by fuxingloh — CHANGES_REQUESTED** [review]
Need more JSDoc for a better understanding of what each class does.

File: packages/jellyfish-wallet-encrypted/__tests__/scryptsy.test.ts (hunk 6 above)
Comment (ivan-zynesis) [reply packages/jellyfish-wallet-encrypted/__tests__/scryptsy.test.ts:36]: Using original suggested from bip0038, I think it takes 1-2 seconds to encrypt or decrypt

File: packages/jellyfish-wallet-encrypted/package.json [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    "@defichain/jellyfish-wallet": "0.0.0",
+    "@defichain/jellyfish-wallet-mnemonic": "0.0.0",
+    "bignumber.js": "^9.0.1",
+    "scryptsy": "^2.1.0",
+    "tiny-secp256k1": "^1.1.6"
+  },
+  "devDependencies": {
+    "@types/scryptsy": "^2.0.0",
+    "@types/tiny-secp256k1": "^1.0.0"
+  }
```
[the remark this answers is not in the document]
Comment (ivan-zynesis) [reply packages/jellyfish-wallet-encrypted/package.json:48]: Updated, 344 will add back some later.

File: packages/jellyfish-wallet-encrypted/__tests__/scryptsy.test.ts (hunk 6 above)
Comment (ivan-zynesis) [reply packages/jellyfish-wallet-encrypted/__tests__/scryptsy.test.ts:36]: I think can be 3-4 seconds depend on how powerful is the phone.

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**